### ISSUE TITLE
support Accept header for "versioning"

### ISSUE THREAD

**Issue by PavelPolyakov** [delvedor/find-my-way#106]
Hi @delvedor,

That's Pavel who contributed to `fastify-swagger` (if that's somehow important ;)).

I'd like to extend `find-my-way` to support "versioning" through the `Accept` header.

Here is some info:
http://labs.qandidate.com/blog/2014/10/16/using-the-accept-header-to-version-your-api/
https://youtu.be/mZ8_QgJ5mbs?t=1275 (starting from 21:15)

I don't know if I succeed, but I want to try. At the end I want to have this available in `fastify`.

Main question to you as maintainer of `find-my-way` at this stage - is this something you may merge in? Or, maybe, for you this has no sense at all.

Merry Christmas 🎅 and looking forward to your response.

Regards,

**Comment by delvedor** [comment]
Hello!
Probably I'm missing something, but versioned routes are supported both in Fastify and this module :)

https://github.com/delvedor/find-my-way#versioned-routes
https://www.fastify.io/docs/latest/Routes/#version

**Comment by PavelPolyakov** [comment]
You are right this is possible, but here are the differences:
1. What is currently supported is semver through `Accept-Version` header (and only)
2. What I want to have supported is
  `Accept: application/vnd.example.api+json;version=2` will lead to one handler
  `Accept: application/vnd.example.api+json;version=3` will lead to the different one
  The actual value of the header may vary, but it's expected that it plays well with the HTTP protocol `Accept` header definition.
  
  There is another concept of versioning, where `Accept` header (which is defined in the http protocol) is used. Basically, in the request we tell the server which media type client may accept and based on that server returns that particular media type.

Yes, this concept is somehow similar with `Accept-Version`, but, we may say that it is more based on the existing http protocol definitions.

**Comment by delvedor** [comment]
I fear that the implementation that you are proposing will affect too much the performances of the router, but if you can work on an implementation that **does not impact** on the current performances, I'm happy to review them.

What I would suggest you is to build a wrapper of the router to handle your scenario, while if you are working with Fastify, the `onRequest` hook should do the work that you need *(especially in Fastify v2, where you can access the context of the route)*.

A Fastify example:
```js
fastify.addHook('onRequest', async (req, reply) => {
  const { accept } = reply.context.config
  yourAcceptLogic(accept, req.headers.accept)
})

fastify.route({
  method: 'GET',
  path: '/',
  config: {
    accept: {
      version: 2,
      type: 'json'
    }
  },
  handler: () => {}
})
```

**Comment by PavelPolyakov** [comment]
Thanks,

Let's see what I can do. You're talking about the results of the `npm run bench`, are you?

> What I would suggest you is to build a wrapper of the router to handle your scenario, while if you are working with Fastify, the onRequest hook should do the work that you need (especially in Fastify v2, where you can access the context of the route).

If I got you right this would mean defining all the handlers logic in `onRequest`. Which is definitely possible, but not that pretty comparing to the standard routes.

**Comment by delvedor** [comment]
Closed in #107.
